File: python/agentsentinel/integrations/openai_assistants.py
"""OpenAI Assistants API integration for AgentSentinel.

Wraps OpenAI's Assistants API function calling with AgentSentinel protection.

Usage::

    from openai import OpenAI
    from agentsentinel import AgentPolicy, AgentGuard
    from agentsentinel.integrations.openai_assistants import (
        OpenAIAssistantsGuard,
        protect_function_map,
    )

    client = OpenAI()
    policy = AgentPolicy(daily_budget=10.0, require_approval=["send_email"])
    guard = AgentGuard(policy)

    # Define your functions
    def get_weather(location: str) -> str:
        return f"Weather in {location}: Sunny, 72°F"

    def send_email(to: str, subject: str, body: str) -> str:
        # This will require approval due to policy
        return f"Email sent to {to}"

    # Protect the function map
    protected_functions = protect_function_map(
        {"get_weather": get_weather, "send_email": send_email},
        guard=guard,
    )

    # Use in your assistant run loop
    if run.required_action:
        tool_outputs = []
        for tool_call in run.required_action.submit_tool_outputs.tool_calls:
            fn_name = tool_call.function.name
            fn_args = json.loads(tool_call.function.arguments)

            # This call is now protected by AgentSentinel
            result = protected_functions[fn_name](**fn_args)
            tool_outputs.append({"tool_call_id": tool_call.id, "output": result})

.. note::
    This module does **not** import ``openai`` at the module level, so
    ``agentsentinel`` has no hard dependency on the OpenAI SDK.
"""

from __future__ import annotations

import json
from typing import Any, Callable, Dict, List, Optional, TypeVar

from ..guard import AgentGuard
from ..policy import AgentPolicy

F = TypeVar("F", bound=Callable[..., Any])


# ---------------------------------------------------------------------------
# Core guard class
# ---------------------------------------------------------------------------


class OpenAIAssistantsGuard:
    """Wraps OpenAI Assistants API function calls with :class:`.AgentGuard`.

    Parameters
    ----------
    guard:
        A pre-configured :class:`.AgentGuard` instance.
    policy:
        :class:`.AgentPolicy` used to create a guard when *guard* is ``None``.
        Ignored when *guard* is provided.
    default_model:
        Default model name used for cost tracking when not overridden per function.

    Example
    -------
    ::

        from agentsentinel import AgentGuard, AgentPolicy
        from agentsentinel.integrations.openai_assistants import OpenAIAssistantsGuard

        policy = AgentPolicy(daily_budget=20.0, model_budgets={"gpt-4o": 10.0})
        oai_guard = OpenAIAssistantsGuard(AgentGuard(policy), default_model="gpt-4o")

        protected = oai_guard.protect_function_map(
            {"get_weather": get_weather, "send_email": send_email}
        )
    """

    def __init__(
        self,
        guard: Optional[AgentGuard] = None,
        policy: Optional[AgentPolicy] = None,
        default_model: str = "gpt-4o",
    ) -> None:
        if guard is not None:
            self._guard = guard
        elif policy is not None:
            self._guard = AgentGuard(policy)
        else:
            self._guard = AgentGuard(AgentPolicy())
        self._default_model = default_model

    @property
    def guard(self) -> AgentGuard:
        """Access the underlying :class:`.AgentGuard`."""
        return self._guard

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def protect_function(
        self,
        func: F,
        name: Optional[str] = None,
        cost: Optional[float] = None,
        model: Optional[str] = None,
    ) -> F:
        """Protect a single function for use with the Assistants API.

        Parameters
        ----------
        func:
            The callable to protect.
        name:
            Override the tool name (defaults to ``func.__name__``).
        cost:
            Explicit cost per invocation (USD).
        model:
            Model name for per-model budget tracking.

        Returns
        -------
        Protected callable with the same signature.
        """
        tool_name = name or func.__name__
        model_name = model or self._default_model
        return self._guard.protect(func, tool_name=tool_name, cost=cost, model=model_name)  # type: ignore[return-value]

    def protect_function_map(
        self,
        functions: Dict[str, Callable[..., Any]],
        costs: Optional[Dict[str, float]] = None,
        models: Optional[Dict[str, str]] = None,
    ) -> Dict[str, Callable[..., Any]]:
        """Return a new dict with every callable protected.

        Parameters
        ----------
        functions:
            Mapping of function names to callables.
        costs:
            Optional per-function cost overrides ``{name: cost_usd}``.
        models:
            Optional per-function model name overrides.

        Returns
        -------
        A **new** dict with the same keys but guarded callables as values.
        """
        costs = costs or {}
        models = models or {}

        protected: Dict[str, Callable[..., Any]] = {}
        for fn_name, func in functions.items():
            protected[fn_name] = self.protect_function(
                func,
                name=fn_name,
                cost=costs.get(fn_name),
                model=models.get(fn_name, self._default_model),
            )
        return protected

    def handle_tool_calls(
        self,
        tool_calls: List[Any],
        functions: Dict[str, Callable[..., Any]],
    ) -> List[Dict[str, str]]:
        """Handle tool calls from an Assistant run, returning tool outputs.

        Parameters
        ----------
        tool_calls:
            List of tool call objects from ``run.required_action.submit_tool_outputs.tool_calls``.
        functions:
            Protected function map (output of :meth:`protect_function_map`).

        Returns
        -------
        List of ``{"tool_call_id": ..., "output": ...}`` dicts ready for
        ``client.beta.threads.runs.submit_tool_outputs``.
        """
        outputs: List[Dict[str, str]] = []
        for tool_call in tool_calls:
            fn_name = tool_call.function.name
            fn_args = json.loads(tool_call.function.arguments)

            try:
                if fn_name in functions:
                    result = functions[fn_name](**fn_args)
                else:
                    result = f"Error: Unknown function '{fn_name}'"
            except Exception as exc:
                result = f"Error: {exc}"

            outputs.append({
                "tool_call_id": tool_call.id,
                "output": str(result),
            })

        return outputs


# ---------------------------------------------------------------------------
# Convenience one-liner
# ---------------------------------------------------------------------------


def protect_function_map(
    functions: Dict[str, Callable[..., Any]],
    *,
    guard: Optional[AgentGuard] = None,
    policy: Optional[AgentPolicy] = None,
    costs: Optional[Dict[str, float]] = None,
    models: Optional[Dict[str, str]] = None,
    default_model: str = "gpt-4o",
) -> Dict[str, Callable[..., Any]]:
    """One-liner: protect every function in *functions* for the Assistants API.

    Parameters
    ----------
    functions:
        Mapping of function names to callables.
    guard:
        Pre-configured :class:`.AgentGuard`.  If ``None``, *policy* must be
        supplied and a new guard is created automatically.
    policy:
        :class:`.AgentPolicy` used to create a guard when *guard* is ``None``.
        Ignored when *guard* is provided.
    costs:
        Optional per-function cost overrides.
    models:
        Optional per-function model name overrides.
    default_model:
        Default model for cost tracking.

    Returns
    -------
    A **new** dict with the same keys but every callable wrapped.

    Raises
    ------
    ValueError:
        When neither *guard* nor *policy* is supplied.

    Example
    -------
    ::

        from agentsentinel import AgentPolicy
        from agentsentinel.integrations.openai_assistants import protect_function_map

        protected = protect_function_map(
            {"get_weather": get_weather, "send_email": send_email},
            policy=AgentPolicy(daily_budget=10.0, require_approval=["send_email"]),
        )
    """
    if guard is None and policy is None:
        raise ValueError("Supply either 'guard' or 'policy'.")

    oai_guard = OpenAIAssistantsGuard(guard=guard, policy=policy, default_model=default_model)
    return oai_guard.protect_function_map(functions, costs=costs, models=models)

File: typescript/src/integrations/openaiAssistants.ts
/**
 * OpenAI Assistants API integration for AgentSentinel.
 *
 * Wraps OpenAI Assistants API function calling with AgentSentinel protection.
 *
 * @example
 * ```ts
 * import { AgentPolicy, AgentGuard } from "@agentsentinel/sdk";
 * import { protectFunctionMap } from "@agentsentinel/sdk/integrations/openaiAssistants";
 *
 * const policy = new AgentPolicy({ dailyBudget: 10.0, requireApproval: ["send_email"] });
 * const guard = new AgentGuard({ policy });
 *
 * const protectedFns = protectFunctionMap(
 *   { get_weather: getWeather, send_email: sendEmail },
 *   guard,
 * );
 * ```
 */

import { AgentGuard } from "../guard";
import { AgentPolicy } from "../policy";

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AnyFn = (...args: any[]) => any;

export interface FunctionMapOptions {
  /** Per-function cost overrides (function name → cost in USD). */
  costs?: Record<string, number>;
  /** Per-function model name overrides. */
  models?: Record<string, string>;
  /** Default model for cost tracking. */
  defaultModel?: string;
}

/**
 * Wraps OpenAI Assistants API function calls with {@link AgentGuard}.
 */
export class OpenAIAssistantsGuard {
  private readonly _guard: AgentGuard;
  private readonly _defaultModel: string;

  constructor(
    guardOrPolicy: AgentGuard | AgentPolicy,
    defaultModel = "gpt-4o",
  ) {
    this._guard =
      guardOrPolicy instanceof AgentGuard
        ? guardOrPolicy
        : new AgentGuard({ policy: guardOrPolicy });
    this._defaultModel = defaultModel;
  }

  /** Access the underlying {@link AgentGuard}. */
  get guard(): AgentGuard {
    return this._guard;
  }

  /**
   * Protect a single function.
   *
   * @param fn     The callable to protect.
   * @param name   Override the tool name (defaults to `fn.name`).
   * @param cost   Explicit cost per invocation (USD).
   * @param model  Model name for per-model budget tracking.
   * @returns      Protected callable with the same signature.
   */
  protectFunction<T extends AnyFn>(
    fn: T,
    name?: string,
    cost?: number,
    model?: string,
  ): T {
    return this._guard.protect(fn, {
      toolName: name ?? fn.name,
      cost,
      model: model ?? this._defaultModel,
    });
  }

  /**
   * Protect a dictionary of functions.
   *
   * @param functions  Mapping of function names to callables.
   * @param opts       Optional cost/model overrides and default model.
   * @returns          New map with the same keys but guarded callables.
   */
  protectFunctionMap(
    functions: Record<string, AnyFn>,
    opts: FunctionMapOptions = {},
  ): Record<string, AnyFn> {
    const costs = opts.costs ?? {};
    const models = opts.models ?? {};
    const defaultModel = opts.defaultModel ?? this._defaultModel;

    const protected_: Record<string, AnyFn> = {};
    for (const [fnName, fn] of Object.entries(functions)) {
      protected_[fnName] = this.protectFunction(
        fn,
        fnName,
        costs[fnName],
        models[fnName] ?? defaultModel,
      );
    }
    return protected_;
  }

  /**
   * Handle tool calls from an Assistant run, returning tool outputs.
   *
   * @param toolCalls  Tool call objects from `run.required_action.submit_tool_outputs.tool_calls`.
   * @param functions  Protected function map.
   * @returns          Array of `{ tool_call_id, output }` objects.
   */
  handleToolCalls(
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    toolCalls: any[],
    functions: Record<string, AnyFn>,
  ): Array<{ tool_call_id: string; output: string }> {
    return toolCalls.map((toolCall) => {
      const fnName: string = toolCall.function.name;
      const fnArgs = JSON.parse(toolCall.function.arguments as string) as Record<string, unknown>;

      let result: string;
      try {
        if (fnName in functions) {
          result = String(functions[fnName](fnArgs));
        } else {
          result = `Error: Unknown function '${fnName}'`;
        }
      } catch (err) {
        result = `Error: ${err instanceof Error ? err.message : String(err)}`;
      }

      return { tool_call_id: toolCall.id as string, output: result };
    });
  }
}

/**
 * One-liner: protect every function in `functions` for the Assistants API.
 *
 * @param functions      Mapping of function names to callables.
 * @param guardOrPolicy  {@link AgentGuard} or {@link AgentPolicy} to use.
 * @param opts           Optional cost/model overrides.
 * @returns              New map with the same keys but every callable wrapped.
 */
export function protectFunctionMap(
  functions: Record<string, AnyFn>,
  guardOrPolicy: AgentGuard | AgentPolicy,
  opts: FunctionMapOptions = {},
): Record<string, AnyFn> {
  const oaiGuard = new OpenAIAssistantsGuard(guardOrPolicy, opts.defaultModel);
  return oaiGuard.protectFunctionMap(functions, opts);
}
